packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -0,0 +1,438 @@
+        // remove the vref now, but some deleteVirtualObject might
+        // shake it loose again for a future pass to investigate
+        possiblyDeadSet.delete(vrefOrBaseRef);
+
+        const parsed = parseVatSlot(vrefOrBaseRef);
+        assert.equal(parsed.type, 'object', vrefOrBaseRef);
+
+        let res = {};
+        if (parsed.virtual || parsed.durable) {
+          const baseRef = vrefOrBaseRef;
+          res = checkExportRepresentative(baseRef);
+        } else if (parsed.allocatedByVat) {
+          const vref = vrefOrBaseRef;
+          res = checkExportRemotable(vref);
+        } else {
+          const vref = vrefOrBaseRef;
+          res = checkImportPresence(vref);
+        }
+
+        // prepare our end-of-crank syscalls
+        if (res.dropImport) {
+          importsToDrop.add(res.dropImport);
+          possiblyRetiredSet.add(res.dropImport);
+        }
+        for (const facetVref of res.exportsToRetire || []) {
+          exportsToRetire.add(facetVref);
+        }
+        gcAgain ||= !!res.gcAgain;
+      }
+
+      // Deleting virtual object state, or freeing weak-keyed
+      // collection entries, might shake loose more
+      // objects. possiblyDeadSet and possiblyRetiredSet are added
+      // when vdata vrefs are decreffed to zero, and gcAgain means that
+      // something in RAM might now be free. In both cases we should
+      // do another pass, including gcAndFinalize(), until we've
+      // cleared everything we can.
+    } while (possiblyDeadSet.size > 0 || gcAgain);
+
+    // Now we process potential retirements. addToPossiblyRetiredSet()
+    // is given vrefs (of all sorts) for the keys of deleted weak
+    // collections, plus any o-NN "Presence-type" vrefs that were
+    // dropped above. We only care about Presence-type, because we
+    // must do retireImports for them: the kernel doesn't care if/when
+    // we stop recognizing our own (maybe-exported) Remotable- and
+    // Representative- type vrefs.
+
+    for (const vref of [...possiblyRetiredSet].sort()) {
+      possiblyRetiredSet.delete(vref);
+      const parsed = parseVatSlot(vref);
+      assert.equal(parsed.type, 'object', vref);
+      // not allocated by us? that makes it a Presence-type
+      if (!parsed.allocatedByVat) {
+        // if we're dropping the vref, checkImportPresence() already
+        // did our isReachable check, so we can safely skip it (and
+        // save a vatstore syscall)
+        const isReachable =
+          !importsToDrop.has(vref) &&
+          (slotToVal.has(vref) || vrm.isPresenceReachable(vref));
+        // note: the old version used getValForSlot, which treats
+        // COLLECTED as unreachable, which would allow a retirement at
+        // the wrong time (before the finalizer fires and does the
+        // dropImport)
+        const isRecognizable = isReachable || vrm.isVrefRecognizable(vref);
+        if (!isRecognizable) {
+          importsToRetire.add(vref);
+        }
+      }
```

Comment:
Reducing indentation depth:
```suggestion
      if (parsed.allocatedByVat) continue;

      // If we're dropping the vref, checkImportPresence() already
      // did our isReachable check, so we can safely skip it (and
      // save a vatstore syscall).
      const isReachable =
        !importsToDrop.has(vref) &&
        // Use slotToVal.has, not getValForSlot(), to avoid retirement
        // before the finalizer fires and does dropImport.
        (slotToVal.has(vref) || vrm.isPresenceReachable(vref));
      const isRecognizable = isReachable || vrm.isVrefRecognizable(vref);
      if (!isRecognizable) {
        importsToRetire.add(vref);
      }
```

Reply:
am I allowed to omit braces on an `if`? I thought that would cause the world to end or something. :)

Reply:
I'm slightly weirded out by splitting an expression with a comment, but I'm living dangerously today, I'll give it a whirl